R/variance_ratio.R

```diff
@@ -33,100 +33,162 @@
+                          species.var = species.var,
+                          abundance.var = abundance.var,
+                          method = variance_ratio_matrixdata,
+                          bootnumber = bootnumber)
+
+      ## confidence intervals
+      null_output <- lapply(null_list, `[[`, i = "out")
+
+      null_means <- vector(mode = "numeric", length = bootnumber)
+
+      for (B in 1:bootnumber) {
+        null_means[B] <- mean(vapply(null_output, `[[`, 1, i = B))
+      }
+
+      li <- (1 - level)/2
+      ui <- 1 - li
+      lowerCI <- quantile(null_means, li)
+      upperCI <- quantile(null_means, ui)
+      nullmean <- mean(null_means)
+      nullout <- data.frame(lowerCI = lowerCI,
+                            upperCI = upperCI,
+                            nullmean =  nullmean)
+
+      # bind actual value and CI
+      output <- cbind(nullout, VR)
+
     } else {
-        
-      # if multiple replicates, check all replicates have values
-        check_single(df, time.var, species.var, replicate.var)
-        df[replicate.var] <- if(is.factor(df[[replicate.var]]) == TRUE) {
-            factor(df[[replicate.var]])
-        } else {
-            df[replicate.var]
-        }
-        
-        # calculate average variance ratio across replicates
-        if(average.replicates == TRUE) {
-           check_multispp(df, species.var, replicate.var)
-            df <- df[order(df[[replicate.var]]),]
-            X <- split(df, df[replicate.var])
-            VR <- mean(unlist(lapply(X, FUN = variance_ratio_longformdata, time.var, species.var, abundance.var))) 
-        } else {
-          
-          # calculate the variance ratio for replicate
-            check_multispp(df, species.var, replicate.var)
-            df <- df[order(df[[replicate.var]]),]
-            X <- split(df, df[replicate.var])
-            VR <- do.call("rbind", lapply(X, FUN = variance_ratio_longformdata, time.var, species.var, abundance.var))
-        }
+
+      # calculate the variance ratio for replicate
+      check_multispp(df, species.var, replicate.var)
+      df <- df[order(df[[replicate.var]]),]
+      X <- split(df, df[replicate.var])
+
+      ## also do separate null models (ie on split list)
+      null_list <- lapply(X, FUN = cyclic_shift,  time.var = time.var,
+                          species.var = species.var,
+                          abundance.var = abundance.var,
+                          method = variance_ratio_matrixdata,
+                          bootnumber = bootnumber)
+      ## calculate confidence intervals for each:
+      null_intervals <- lapply(null_list, confint)
+
+      ## combine as data.frames and preserve replicate var
+      nullout <- purrr::map_df(null_intervals, as.data.frame, .id = replicate.var)
+
+      VR_list <- lapply(X, FUN = variance_ratio_longformdata,
+                        time.var, species.var, abundance.var)
+
+      VRout <- purrr::map_df(VR_list, ~data.frame(VR = .), .id = replicate.var)
```

Comment:
Oh cool I didn't know about that package! But since I don't know it, let's
check with Matt before building the dependency

On Fri, Mar 18, 2016 at 1:06 AM, Andrew MacDonald <notifications@github.com>
wrote:

> In R/variance_ratio.R
> <https://github.com/laurenmh/codyn/pull/67#discussion_r56619448>:
>
> > +      ## also do separate null models (ie on split list)
> > +      null_list <- lapply(X, FUN = cyclic_shift,  time.var = time.var,
> > +                          species.var = species.var,
> > +                          abundance.var = abundance.var,
> > +                          method = variance_ratio_matrixdata,
> > +                          bootnumber = bootnumber)
> > +      ## calculate confidence intervals for each:
> > +      null_intervals <- lapply(null_list, confint)
> > +
> > +      ## combine as data.frames and preserve replicate var
> > +      nullout <- purrr::map_df(null_intervals, as.data.frame, .id = replicate.var)
> > +
> > +      VR_list <- lapply(X, FUN = variance_ratio_longformdata,
> > +                        time.var, species.var, abundance.var)
> > +
> > +      VRout <- purrr::map_df(VR_list, ~data.frame(VR = .), .id = replicate.var)
>
> here are 2/3 of the places where I lean on purrr::map_df. I was running
> out of time and this was the best I could think of to solve this problem. I
> like that it let's us recover the name of replicate.var. What do you
> think? I know we are striving for minimal dependencies
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/laurenmh/codyn/pull/67/files/f3e38639882cb3e07b489b62de4bd6214853b215#r56619448>
>